Issues/370

ravinderk opened the pull request:
## Description
<!-- Please describe your changes -->
This Pr will resolve #370 

Note:
1. Any address coming from the donation will be considered as a billing address.
2. Any address coming from profile shortcode will be considered as a personal address.
2. All address field meta keys contain `line1`, `line2`, 'city', `state`, `country', and `zip` in there meta key name.

## How Has This Been Tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
1. Billing address auto-fill for logged in the donor.
2. Profile shortcode.
3. Address related actions on donor screen in admin.
4. Processing donations.
5. Running upgrades.


## Screenshots (jpeg or gifs if applicable):
![ok0wqebqwm 1](https://user-images.githubusercontent.com/1784821/31952550-80283bb2-b8fe-11e7-98c5-f22bb7260f4a.gif)

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

ravinderk commented on assets/js/admin/admin-scripts.js:
```diff
@@ -12,11 +12,48 @@ jQuery.noConflict();
+	 * @param $parent
+	 * @param args
+	 */
+	var giveAjaxLoader = function ( $parent, args ){
```
@DevinWalker New function will handle ajax show/hide action.

ravinderk commented on assets/js/plugins/give-ajaxify-fields.js:
```diff
@@ -0,0 +1,120 @@
+	'use strict';
+
+	$.extend({
+		giveAjaxifyFields: function( customSettings ) {
```
@DevinWalker New plugin to handle ajaify fields like the country select option. currently, it supports only country state field

ravinderk commented on assets/scss/admin/util.scss:
```diff
@@ -39,4 +39,37 @@
+/**
+ * Spinner
+ */
+.give-spinner-wrapper{
```
@DevinWalker New spinner related classed which we can use in core

ravinderk commented:
@DevinWalker All test is passing on my local. I restarted test, let me check if it is related to pr or not.

<img width="703" alt="screen shot 2017-10-25 at 8 34 53 am" src="https://user-images.githubusercontent.com/1784821/31978620-d80f1150-b95f-11e7-83fa-c0c5b7fe5695.png">

DevinWalker commented:
@ravinderk I was able to reproduce the error here: https://travis-ci.org/WordImpress/Give/jobs/292417738 locally:

![2017-10-25_19-31-26](https://user-images.githubusercontent.com/1571635/32032277-215cfb6e-b9bb-11e7-8922-2ef38ce87e71.png)

DevinWalker commented:
## Review Findings

**1. No addresses:** A message should display.
![2017-10-25_19-33-12](https://user-images.githubusercontent.com/1571635/32032351-8c481954-b9bb-11e7-9621-311696e180e9.png)

**2. Swap City and State:** The city and state fields need to be swapped. Related to https://github.com/WordImpress/Give/issues/2226
![2017-10-25_19-35-45](https://user-images.githubusercontent.com/1571635/32032405-dbe8407e-b9bb-11e7-861c-11b27975946e.png)

**3. Adding Address Didn't Work:** I attempted to add an address to this donor and it didn't allow me:
![2017-10-25_19-39-25](https://user-images.githubusercontent.com/1571635/32032487-57aa1462-b9bc-11e7-8029-08619952ee40.gif)

**4. Message Incorrect:** This should read better:
![2017-10-25_19-41-04](https://user-images.githubusercontent.com/1571635/32032507-761326e6-b9bc-11e7-9d6b-370831cbc2f8.png)

> **Error:** Unable to save the address. Please check if the address already exists. 

"Check after sometime" brings zero value and only confuses. 

**5. Numbering:** Numbering should always start at `1`, not `0`: 
![2017-10-25_19-42-09](https://user-images.githubusercontent.com/1571635/32032557-af9930c2-b9bc-11e7-9ac6-84e924ab3ce9.png)

**6. Spacing:** There is a significant gap at the bottom of the address that should be tightened up. Provide spacing for up to **6 max lines** per address. Remove the extra whitespace.
![2017-10-25_19-43-34](https://user-images.githubusercontent.com/1571635/32032615-e65d30c2-b9bc-11e7-9f0c-352bd9246847.png)

**7. Added Address Success Message:** When an address is successfully added a message should display within the address fieldset:
![2017-10-25_19-47-55](https://user-images.githubusercontent.com/1571635/32032714-6fd67d22-b9bd-11e7-8b79-344be561bf09.png)

**8. Move "Add Address Button Below Addresses**: It's too in the way where it's at and people will be more interested in viewing the data than seeing the button.
![2017-10-25_19-49-04](https://user-images.githubusercontent.com/1571635/32032740-934e820e-b9bd-11e7-90a3-bc4058aac09f.png)

DevinWalker commented on includes/admin/admin-actions.php:
```diff
@@ -630,4 +630,182 @@ function give_get_user_roles() {
+				'error' => 1,
+				'error_msg' => wp_sprintf(
+					'<div class="notice notice-error"><p>%s</p></div>',
+					__( 'Error: Security issue. please try after sometime.', 'give' )
```
Remove "Please try after some time" that doesn't make sense and only frustrates users. Trying after sometime won't change the end result.

DevinWalker commented on includes/admin/admin-actions.php:
```diff
@@ -630,4 +630,182 @@ function give_get_user_roles() {
+						'error' => 1,
+						'error_msg' => wp_sprintf(
+							'<div class="notice notice-error"><p>%s</p></div>',
+							__( 'Error: could not able to save address. check if address already exist or not otherwise please try after sometime.', 'give' )
```
Already provided an update for this text within the review on PR comment.

DevinWalker commented on includes/admin/admin-actions.php:
```diff
@@ -630,4 +630,182 @@ function give_get_user_roles() {
+						'error' => 2,
+						'error_msg' => wp_sprintf(
+							'<div class="notice notice-error"><p>%s</p></div>',
+							__( 'Error: could not able to delete address. please try after sometime.', 'give' )
```
Same here as above with the "try after sometime" sentence.

DevinWalker commented on includes/admin/admin-actions.php:
```diff
@@ -630,4 +630,182 @@ function give_get_user_roles() {
+						'error' => 3,
+						'error_msg' => wp_sprintf(
+							'<div class="notice notice-error"><p>%s</p></div>',
+							__( 'Error: could not able to update address. check if address already exist or not otherwise please try after sometime.', 'give' )
```
Better message provided for this in PR comments.

ravinderk commented:
@DevinWalker PR updated.

**3. Adding Address Didn't Work:** previously I removed validation from state field as some country does not have this other was required. To resolve this I removed validation from `Address 2` also. Address form does not have validation like frontend billing address, we will add it if necessary.

ravinderk commented on includes/misc-functions.php:
```diff
@@ -1610,11 +1610,18 @@ function give_delete_meta( $id, $meta_key, $meta_value = '' ) {
 
+	// Fresh install?
+	// If fresh install then all upgrades will be consider as completed.
+	if ( ! get_option( 'give_version' ) ) {
```
@DevinWalker On fresh install if anybody uses it before plugin activates hook then it will return false but for the fresh install, all upgrades will complete by default, so I updated this function.

DevinWalker replied:
Using `get_option` within a conditional will cause a PHP fatal write error in old versions of PHP. You have to save it to a variable before.

ravinderk replied:
Fixed https://github.com/WordImpress/Give/pull/2233/commits/4f52aaa6510999e6c6dfaed790803e28adb3e793

DevinWalker commented:
## Second Review Findings

1. Fixed

2. Fixed

3. Fixed

4. The message is still incorrect:

![2017-10-30_14-02-29](https://user-images.githubusercontent.com/1571635/32195636-15a67e4e-bd7b-11e7-9178-3e92e5488f83.png)

Please see above for how it should read.

5. Fixed

6. The address blocks need a min-height to prevent blocks bunching like this:

![2017-10-30_14-04-55](https://user-images.githubusercontent.com/1571635/32195715-577e5f4e-bd7b-11e7-8e9d-e9f69267bbcd.png)

7. Fixed

8. Fixed

---

Please resolve numbers 4 and 6 above.

ravinderk commented:
@DevinWalker Point 4 and 6 has been resolved.
